File: frontend/src/api/mockExams.ts
import { apiClient } from "./client"
import type {
  MockExam,
  MockExamGenerateResponse,
  MockExamListItem,
  MockExamResults,
  MockExamSubmitResponse,
} from "./types"

export async function generateMockExam(courseId: string): Promise<MockExamGenerateResponse> {
  const response = await apiClient.post<MockExamGenerateResponse>(
    `/courses/${courseId}/mock-exams/generate`,
  )
  return response.data
}

export async function listMockExams(courseId: string): Promise<MockExamListItem[]> {
  const response = await apiClient.get<MockExamListItem[]>(`/courses/${courseId}/mock-exams`)
  return response.data
}

export async function getMockExam(mockExamId: string): Promise<MockExam> {
  const response = await apiClient.get<MockExam>(`/mock-exams/${mockExamId}`)
  return response.data
}

export async function submitMockExam(
  mockExamId: string,
  answers: Array<{
    question_id: string
    student_answer: string
    time_spent_seconds: number
  }>,
): Promise<MockExamSubmitResponse> {
  const response = await apiClient.post<MockExamSubmitResponse>(
    `/mock-exams/${mockExamId}/submit`,
    { answers },
  )
  return response.data
}

export async function getMockExamResults(mockExamId: string): Promise<MockExamResults> {
  const response = await apiClient.get<MockExamResults>(`/mock-exams/${mockExamId}/results`)
  return response.data
}

File: backend/api/routes/mock_exams.py
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase

from api.dependencies import get_current_user
from core.database import get_database
from schemas.mock_exam_schemas import (
    MockExamGenerateResponse,
    MockExamListItem,
    MockExamResponse,
    MockExamResultsResponse,
    MockExamSubmitRequest,
    MockExamSubmitResponse,
)
from services.mock_exam_service import MockExamService

router = APIRouter(tags=["mock-exams"])


@router.post("/courses/{course_id}/mock-exams/generate", response_model=MockExamGenerateResponse)
async def generate_mock_exam(
    course_id: str,
    current_user: dict = Depends(get_current_user),
    database: AsyncIOMotorDatabase = Depends(get_database),
) -> MockExamGenerateResponse:
    return await MockExamService(database).generate_mock_exam(
        str(current_user["_id"]),
        course_id,
        current_user.get("language", "en"),
    )


@router.get("/courses/{course_id}/mock-exams", response_model=list[MockExamListItem])
async def list_mock_exams(
    course_id: str,
    current_user: dict = Depends(get_current_user),
    database: AsyncIOMotorDatabase = Depends(get_database),
) -> list[MockExamListItem]:
    return await MockExamService(database).list_mock_exams(
        str(current_user["_id"]),
        course_id,
    )


@router.get("/mock-exams/{mock_exam_id}", response_model=MockExamResponse)
async def get_mock_exam(
    mock_exam_id: str,
    current_user: dict = Depends(get_current_user),
    database: AsyncIOMotorDatabase = Depends(get_database),
) -> MockExamResponse:
    return await MockExamService(database).get_mock_exam(
        str(current_user["_id"]),
        mock_exam_id,
    )


@router.post("/mock-exams/{mock_exam_id}/submit", response_model=MockExamSubmitResponse)
async def submit_mock_exam(
    mock_exam_id: str,
    payload: MockExamSubmitRequest,
    current_user: dict = Depends(get_current_user),
    database: AsyncIOMotorDatabase = Depends(get_database),
) -> MockExamSubmitResponse:
    return await MockExamService(database).submit_mock_exam(
        str(current_user["_id"]),
        mock_exam_id,
        payload,
        current_user.get("language", "en"),
    )


@router.get("/mock-exams/{mock_exam_id}/results", response_model=MockExamResultsResponse)
async def get_mock_exam_results(
    mock_exam_id: str,
    current_user: dict = Depends(get_current_user),
    database: AsyncIOMotorDatabase = Depends(get_database),
) -> MockExamResultsResponse:
    return await MockExamService(database).get_results(
        str(current_user["_id"]),
        mock_exam_id,
    )
